sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -440,6 +440,44 @@ case class CreateTableAsSelect(
   }
 }
 
+/**
+ * Replace a table with a v2 catalog.
+ *
+ * If the table does not exist, it will be created. The persisted table will have no contents
+ * as a result of this operation.
+ */
+case class ReplaceTable(
+    catalog: TableCatalog,
+    tableName: Identifier,
+    tableSchema: StructType,
+    partitioning: Seq[Transform],
+    properties: Map[String, String],
+    orCreate: Boolean) extends Command
+
+/**
+ * Replaces a table from a select query with a v2 catalog.
+ *
+ * If the table does not already exist, it will be created.
+ */
+case class ReplaceTableAsSelect(
```

Comment:
What is the behavior if `query` depends on the table being replaced?

Reply:
Great question. I'm fairly certain that:
- If non-atomic catalog is being used, the table will have been dropped, so loading the table for the query will result in an error
- If the atomic catalog is being used, the drop part of the replace isn't committed yet, so the catalog should be able to load the table's contents before the write is committed.

@rdblue - curious to hear your thoughts on the first situation. It makes a stronger case for trying to ban replacing tables without an atomic catalog.

Reply:
I agree. If the drop happens first and is not atomic, then the create will fail.

We should be able to add a rule to check for this and fail the query in analysis if the table doesn't support atomic updates. @mccheah, can you open an issue for adding a rule like that?

Reply:
+1, I think we should fail the query if atomic REPLACE TABLE is not supported. It's a valid use case to access the table being replaced in RTAS,  Spark shouldn't throw table not found exception in this case, which is quite confusing.

Reply:
@cloud-fan, we discussed this in the DSv2 sync and decided that, in general, RTAS should be supported even if the source is not atomic. Just to clarify, this would be a rule to fail RTAS if it is not atomic when the query depends on the table being replaced. We don't want to drop a table and then find that we can't replace it because it was dropped.